File: ts_sdk/src/entities/graph-workflow-node.ts
import { APIEntity, registerEntity } from '../APIEntity';
import { IEntity, EntityMerge } from '../IEntity';

/**
 * GraphWorkflowNode — derived record-keeping row for a node inside an GraphWorkflow
 * (backend: flow_sdk/builtin/flow_node.py). The SOURCE OF TRUTH for nodes is
 * the flow's graph.json; these rows exist for ownership/cleanup/queries only.
 */
export interface IGraphWorkflowNode extends IEntity {
  name?: string;
  flow_id?: string;
  node_type?: 'trigger' | 'agent' | 'function';
  program_ref?: string;
  enabled?: boolean;
}

// `implements IGraphWorkflowNode` only checks the class; it contributes no members, so every
// field declared solely on IGraphWorkflowNode read as "does not exist". deepAssign populates
// them from the wire — this merge makes them part of the class type.
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface GraphWorkflowNode extends EntityMerge<IGraphWorkflowNode> {}

@registerEntity
export class GraphWorkflowNode extends APIEntity<GraphWorkflowNode> implements IGraphWorkflowNode {
  name?: string;
  flow_id?: string;
  node_type?: 'trigger' | 'agent' | 'function';
  program_ref?: string;
  enabled?: boolean;
  static type: string = 'graph_workflow_node';

  constructor(entity: Partial<IGraphWorkflowNode> = {}) {
    super(entity);
    this.name = entity.name ?? '';
    this.flow_id = entity.flow_id ?? '';
    this.node_type = entity.node_type ?? 'function';
    this.program_ref = entity.program_ref;
    this.enabled = entity.enabled ?? true;
  }
}

File: flow_sdk/builtin/graph_workflow_node.py
"""GraphWorkflowNode — derived record-keeping row for a node inside an GraphWorkflow.

The SOURCE OF TRUTH for nodes is the flow's ``graph.json`` (see
``flow_sdk/graph_workflow_manager/graph_workflow_doc.py``); routing never reads these rows.
The indexer syncs one row per agent/function node, attached as a CHILD
of the GraphWorkflow — so ownership, cleanup, and "which flows use skill X"
queries work through the ordinary entity graph.

DB-only entity (no ``asset_ref``).
"""
from typing import ClassVar, Optional

from flow_sdk.api.api_types.api_field import APIField
from flow_sdk.core import Entity
from flow_sdk.schema.types import EntityType

# Model size → CLI model alias for spawned agent executions. Size names (not
# model ids) so the mapping can evolve without touching stored flows.
MODEL_SIZE_TO_CLI = {
    "sm": "haiku",
    "md": "sonnet",
    "lg": "opus",
}


class GraphWorkflowNode(Entity):
    type: str = APIField(default=EntityType.GRAPH_WORKFLOW_NODE.value)
    name: str = APIField(default="")
    flow_id: str = APIField(default="", description="The GraphWorkflow this node belongs to.")
    node_type: str = APIField(default="function", description="trigger | agent | function")
    program_ref: Optional[str] = APIField(None, description="Program summary for querying (e.g. skill name).")
    enabled: bool = APIField(default=True)

    _api_visible: ClassVar[bool] = True
